integration/container_stats_test.go

```diff
@@ -69,6 +69,44 @@ func TestContainerStats(t *testing.T) {
 	testStats(t, s, containerConfig)
 }
 
+func TestContainerStatsDoesNotErrorOnStoppingContainer(t *testing.T) {
+	t.Logf("Create a pod config and run sandbox container")
+	sb, sbConfig := PodSandboxConfigWithCleanup(t, "sandbox1", "stats")
+
+	EnsureImageExists(t, pauseImage)
+
+	t.Logf("Create a container config and run container in a pod")
+	containerConfig := ContainerConfig(
+		"container1",
+		pauseImage,
+		WithTestLabels(),
+		WithTestAnnotations(),
+	)
+	cn, err := runtimeService.CreateContainer(sb, containerConfig, sbConfig)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, runtimeService.RemoveContainer(cn))
+	}()
+	require.NoError(t, runtimeService.StartContainer(cn))
+	defer func() {
+		runtimeService.StopContainer(cn, 10)
+	}()
+
+	t.Logf("Verify that there are no errors when collecting stats from a stopping container.")
+	go func() {
+		// The shim will return an empty stats response on a stopped container. We need to check
+		// that we don't fail in cases where the shim might return a nil response. This can happen
+		// if the container is stopping.
+		assert.NoError(t, runtimeService.StopContainer(cn, 10))
+	}()
+
+	for i := 0; i < 50; i++ {
```

Comment:
might be more definitive to check stats before and after stopping the container...

Reply:
This error gets triggered on Windows only when a container is in the process of being stopped. The Windows shim has a number of go routines that wait for certain events and cleanup various resources claimed by the container. If a ```Stats()``` request is received during the cleanup process where the handle to the container is being closed, an empty stats response will be returned:

```bash
&Statistics{Container:<nil>,VM:nil,XXX_unrecognized:[],}
```

The error generated by the shim is:

```ini
time="2022-01-12T04:34:36.965282100-08:00" level=error msg="collecting metrics for 70daae139b18bad4232d2ce7d5f6d193ca72646b6caeb54f46be3b24ef7751e8" error="hcs::System::PropertiesV2 70daae139b18bad4232d2ce7d5f6d193ca72646b6caeb54f46be3b24ef7751e8: The handle is invalid.: unknown"
```

I am now wondering if it wouldn't be better to  change the ```getTasksMetrics()``` function here:

https://github.com/containerd/containerd/blob/main/services/tasks/local.go#L644

and return an error instead of continuing if calling the ```Stats()``` function fails here:

https://github.com/containerd/containerd/blob/main/services/tasks/local.go#L661

What do you think?

Marking this as a [WiP] until we decide what to do. In the meantime, I'll also open a PR against ```hcsshim```.

Reply:
Opened this PR against ```hcsshim```:

https://github.com/microsoft/hcsshim/pull/1276

Which should fix the issue picked up by this test. I will revert the changes in ``` pkg/cri/server/container_stats_list_windows.go```